packages/completer/src/handler.ts

```diff
@@ -434,19 +481,48 @@ export class CompletionHandler implements IDisposable {
 
     // Update the options, including the type map.
     model.setOptions(matches, typeMap);
+  }
 
-    // Update the cursor.
-    model.cursor = {
-      start: Text.charIndexToJsIndex(reply.start, text),
-      end: Text.charIndexToJsIndex(reply.end, text)
-    };
+  /**
+   * Receive completion items from provider.
+   *
+   * @param state - The state of the editor when completion request was made.
+   *
+   * @param reply - The API response returned for a completion request.
+   */
+  private _onFetchItemsReply(
+    state: Completer.ITextState,
+    reply: CompletionHandler.ICompletionItemsReply
+  ) {
+    const model = this._updateModel(state, reply.start, reply.end);
+    if (!model) {
+      return;
+    }
+    if (model.setCompletionItems) {
+      model.setCompletionItems(reply.items);
+    }
   }
 
-  private _connector: IDataConnector<
-    CompletionHandler.IReply,
-    void,
-    CompletionHandler.IRequest
-  >;
+  /**
+   * If completion request fails, reset model and fail silently.
+   */
+  private _onFailure() {
+    const model = this.completer.model;
+
+    if (model) {
+      model.reset(true);
+    }
+  }
+
+  private _connector:
+    | IDataConnector<CompletionHandler.IReply, void, CompletionHandler.IRequest>
+    | (IDataConnector<
+        CompletionHandler.ICompletionItemsReply,
+        void,
+        CompletionHandler.IRequest
+      > & {
+        responseType: typeof CompletionHandler.ICompletionItemsResponseType;
+      });
```

Comment:
We should probably make this a named type since we use it a few places here and in the token.